specutils/analysis/correlation.py

```diff
@@ -0,0 +1,119 @@
+import numpy as np
+
+import astropy.units as u
+from astropy import constants as const
+
+
+from ..manipulation import FluxConservingResampler
+
+
+def _normalize(observed_spectrum, template_spectrum):
+    """
+    Calculate a scale factor to be applied to the template spectrum so the
+    total flux in both spectra will be the same.
+
+    Parameters
+    ----------
+    observed_spectrum : :class:`~specutils.Spectrum1D`
+        The observed spectrum.
+    template_spectrum : :class:`~specutils.Spectrum1D`
+        The template spectrum, which needs to be normalized in order to be
+        compared with the observed spectrum.
+
+    Returns
+    -------
+    `float`
+        A float which will normalize the template spectrum's flux so that it
+        can be compared to the observed spectrum.
+    """
+    num = np.nansum((observed_spectrum.flux*template_spectrum.flux)/
+                 (observed_spectrum.uncertainty.array**2))
+    denom = np.nansum((template_spectrum.flux/
+                    observed_spectrum.uncertainty.array)**2)
+
+    return num/denom
+
+
+def template_correlate(observed_spectrum, template_spectrum):
+    """
+    Compute cross-correlation of the observed and template spectra.
+    This assumes that both are resampled into a common wavelength
+    scale before being submitted to this function.
+
+    Parameters
+    ----------
+    observed_spectrum : :class:`~specutils.Spectrum1D`
+        The observed spectrum.
+    template_spectrum : :class:`~specutils.Spectrum1D`
+        The template spectrum, which will be correlated with
+        the observed spectrum.
+
+    Returns
+    -------
+    (`~astropy.units.Quantity`, `~astropy.units.Quantity`)
+        Arrays with correlation values and lags in km/s
+    """
+    # Normalize template
+    normalization = _normalize(observed_spectrum, template_spectrum)
+
+    # Not sure if we need to actually normalize the template. Depending
+    # on the specific data uncertainty, the normalization factor
+    # may turn out negative. That causes a flip of the correlation function,
+    # in which the maximum (correlation peak) is no longer meaningful.
+    if normalization < 0.:
+        normalization = 1.
+
+    # Resample both spectrum and template to log wavelength. This is assuming
+    # that both spectrum and template share a common spectral axis.
+    resampler = FluxConservingResampler()
```

Comment:
It also may not be essential to conserve flux for this application, since it is just trying to measure the shift. A simpler resampler might be faster.

Reply:
Indeed, easy to add a keyword. As for the simpler resampler, I ended up using the flux conserving one at some point in the development, because it gave better results. But that could have been due to something else as well. Once I got the function+galaxy notebook working, I forgot to revisit the resampler type issue. I will try with the linear type and see how it goes.

Reply:
@hcferguson indeed, replacing the flux conserving by the linear resampler speeded up the calculation significantly. And didn't result in any visible side effects (as we should expect anyways).

Reply:
Note this *did* change the results quite noticeably for the stellar case.  Not in a way that significantly affected the final answer, but in a way I strongly suspect matters a lot more for fainter spectra than my test.  So that just emphasizes the importance of allowing the user to choose "let me do my own resampling, I know what I'm doing" if they want to.